[docs][WIP] Java SDK docs (#43)
dapr-sandbox/components-java-sdk: components-java-sdk/daprdocs/content/en/java-sdk-docs/java-state-store/_index.md

Create the file `components/statestore.java` and add `import` statements for the state store related packages.

```java
package io.dapr.components.wrappers;

import com.google.protobuf.ByteString;
import dapr.proto.components.v1.State;
import dapr.proto.components.v1.State.BulkDeleteRequest;
import dapr.proto.components.v1.State.BulkGetRequest;
import dapr.proto.components.v1.State.BulkGetResponse;
import dapr.proto.components.v1.State.BulkSetRequest;
import dapr.proto.components.v1.State.BulkStateItem;
import dapr.proto.components.v1.State.DeleteRequest;
import dapr.proto.components.v1.State.Etag;
import dapr.proto.components.v1.State.GetRequest;
import dapr.proto.components.v1.StateStoreGrpc;
import io.dapr.components.domain.state.BulkGetError;
import io.dapr.components.domain.state.GetResponse;
import io.dapr.components.domain.state.SetRequest;
import io.dapr.components.domain.state.StateStore;
import io.dapr.v1.ComponentProtos;
import io.dapr.v1.ComponentProtos.FeaturesResponse;
import io.grpc.stub.StreamObserver;
import reactor.core.publisher.Mono;
```

## Implement the `StateStore` interface

Create a class that implements the `StateStore` interface.

```java

> public class StateStoreGrpcComponentWrapper extends StateStoreGrpc.StateStoreImplBase {
> 
>   @Override
>   public void init(final State.InitRequest request, final StreamObserver<State.InitResponse> responseObserver) {
>     // Called to initialize the component with its configured metadata
>   }
> 
>   @Override
>   public void delete(final DeleteRequest request, final StreamObserver<State.DeleteResponse> responseObserver) {
>     // Delete the requested key from the state store
>   }
> 
>   @Override
>   public void get(final GetRequest request, final StreamObserver<State.GetResponse> responseObserver) {
>     // Get the requested key value from the state store
>   }
> 
> 
>   @Override
>   public void set(final State.SetRequest request, final StreamObserver<State.SetResponse> responseObserver) {
>     // Set the requested key to the specified value in the state store
>   }
> 
> 
> 
> }

Reviewer: Let's use the example code we have from https://github.com/dapr-sandbox/components-java-sdk/blob/main/components-java-sdk-examples/src/main/java/io/dapr/examples/statestore/InMemoryStateStore.java 

```suggestion
public class InMemoryStateStore implements StateStore {

  private final Map<String, GetResponse> entries = new HashMap<>();

  @Override
  public Mono<Void> init(Map<String, String> properties) {
    return Mono.empty();
  }

  @Override
  public Mono<GetResponse> get(GetRequest getRequest) {
    return Mono.justOrEmpty(entries.get(getRequest.key()));
  }

  @Override
  public Mono<Void> delete(DeleteRequest deleteRequest) {
    final String key = deleteRequest.key();
    final GetResponse existingData = entries.get(key);
    if (existingData == null) {
      // Key not found: nothing to remove, nothing else to do.
      return Mono.empty();
    } else if (existingData.etag().equals(deleteRequest.etag())) {
      // Valid remove request
      entries.remove(key);
      return Mono.empty();
    } else {
      // Data exists but etag mismatch
      return Mono.error(Exceptions.getEtagMismatchException("in-memory-etag-mismatch-on-delete"));
    }
  }

  @Override
  public Mono<Void> set(SetRequest setRequest) {
    var upsertEntry = new GetResponse(
        setRequest.value(),
        setRequest.etag(),
        setRequest.metadata(),
        setRequest.contentType()
    );
    entries.put(setRequest.key(), upsertEntry);
    // Nothing else to do but return with no errors.
    return Mono.empty();
  }
}
```

```

## Register state store component

In the main application file (for example, `main.java`), register the state store with an application service.

```java

>   public PluggableComponent withStateStore(StateStore stateStore) {
>     Objects.requireNonNull(stateStore);
>     assert !alreadyAddedStateStore; // No, you cannot add multiple stateStores with the same name.
>     alreadyAddedStateStore = true;
> 
>     exposedServices.add(new StateStoreGrpcComponentWrapper(stateStore));
>     // Register other facets of a stateStore, like QueriableStateStore and TransactionalStateStore
>     // if the current stateStore object supports those facets.
>     if (stateStore instanceof QueriableStateStore queriableStore) {
>       exposedServices.add(new QueriableStateStoreComponentWrapper(queriableStore));
>     }
>     if (stateStore instanceof TransactionalStateStore transactionalStateStore) {
>       exposedServices.add(new TransactionalStateStoreComponentWrapper(transactionalStateStore));
>     }
> 
>     return this;
>   }

Reviewer: ```suggestion
    final PluggableComponentServer server = new PluggableComponentServer();
    server.registerComponent(PluggableComponent
            .withName("my-state-store")
            .withStateStore(new StateStoreName())
        )
```

```

## Bulk state stores

> State stores that intend to support bulk operations should implement the optional interfaces. Its methods mirror those of the base `StateStore` interface, but include multiple requested values.

Reviewer: ```suggestion
State stores might also provide support for bulk operations, i.e., operations that handle multiple items through a single request.  A naive implementation is provided by default so that all state stores might support these operations. Under the hood, it uses the single-items operations.

You can customize how bulk operations are handled by overriding the default implementation with your own. This might be necessary if, for instance, bulk operations do not make sense for your specific state store.

Another reason to provide custom bulk operations implementations is to optimize how those are handled, making use of special APIs or native support for bulk operations provided by the underlying database or storage system your are exposing through your pluggable component. 
```

```java

>   @Override
>   public void bulkDelete(final BulkDeleteRequest request,
>                          final StreamObserver<State.BulkDeleteResponse> responseObserver) {
>     // Delete all of the requested values from the state store
>   }
> 
>   @Override
>   public void bulkGet(final BulkGetRequest request,
>                       final StreamObserver<BulkGetResponse> responseObserver) {
>     // Return the values of all of the requested values from the state store
>   }
> 
>   @Override
>   public void bulkSet(final BulkSetRequest request, final StreamObserver<State.BulkSetResponse> responseObserver) {
>     // Set all of the valeus of the requested keys in the state store
>   }

Reviewer: ```suggestion
  @Override
  public Flux<BulkGetStateItem> bulkGet(List<GetRequest> getRequests) {
     // your code here
  }

  public Mono<Void> bulkDelete(List<DeleteRequest> deleteRequests) {
     // your code here
  }

  public Mono<Void> bulkSet(List<SetRequest> setRequests) {
     // your code here
  }
}
```

```

## Transactional state stores

State stores that intend to support transactions should implement the optional `TransactionalStateStore` interface. Its `transact()` method is passed a request with a sequence of delete and/or set operations to be performed within a transaction. 

```java
```

## Queryable state stores

State stores that intend to support queries should implement the optional `QueriableStateStore` interface. `query()` is passed details about the query, such as the filter(s), result limits, pagination, and sort order(s) of the results. The state store uses those details to generate a set of values to return as part of its response.

```java

> public class QueriableStateStoreComponentWrapper extends QueriableStateStoreGrpc.QueriableStateStoreImplBase {
> 
>   // ...
> 
>   @Override
>   public void query(State.QueryRequest request, StreamObserver<State.QueryResponse> responseObserver) {
>     // Generate and return results
>   }
> }

Reviewer: ```suggestion
public class MyQueriableStateStore implements QueriableState, StateStore {

  // StateStore interface methods... 

  // QueriableStateStore requires implementation of the following method:
  public Mono<QueryResponse> query(QueryRequest request) {
     // Parses the query, executes it and creates a QueryResponse object to return.
     return response;
  }
}
```
